14/umbraco-cms/fundamentals/backoffice/property-editors/built-in-umbraco-property-editors/content-picker.md

```diff
@@ -1,41 +1,195 @@
+
+The `ExecuteAsync` method gets a collection of content keys from the last executed query step or the origin. It has to return a new collection of content keys.
+
+```csharp
+public class MyCustomDynamicRootQueryStep : IDynamicRootQueryStep
+{
+    private readonly IMyCustomRepository _myCustomRepository;
+
+    public MyCustomDynamicRootQueryStep(IMyCustomRepository myCustomRepository)
+    {
+        _myCustomRepository = myCustomRepository;
+    }
+
+    public async Task<Attempt<ICollection<Guid>>> ExecuteAsync(ICollection<Guid> origins, DynamicRootQueryStep filter)
+    {
+        if (filter.Alias != "MyCustom") // This is the string you will have to write in the UI
+        {
+            return Attempt<ICollection<Guid>>.Fail();
+        }
+
+        if (origins.Any() is false)
+        {
+            return Attempt<ICollection<Guid>>.Succeed(Array.Empty<Guid>());
+        }
+
+        // TODO replace with your own logic
+        var result = await _myCustomRepository.GetWhateverIWantAsync(origins);
+
+        return Attempt<ICollection<Guid>>.Succeed(result);
+    }
+}
+```
+
+### Filter out items with type
+
+Allow or disallow tree nodes with a certain content type alias.
+
+Enter `typeAlias,altTypeAlias` to only allow selecting nodes with those alias'. Enter `!typeAlias,altTypeAlias` to only allow selecting nodes **not** with those alias'.
+
+### Minimum/maximum number of items
+
+Set a limit on the number of items allowed to be selected.
 
 ## Data Type Definition Example
 
-![Content Picker Data Type Definition](images/Content-Picker2-DataType.png)
+![Content Picker Data Type Definition](../../images/mntp.png)
 
 ## Content Example
 
-![Content Picker Content](../../../../../../10/umbraco-cms/fundamentals/backoffice/property-editors/built-in-property-editors/images/Content-Picker-Content-v10.png)
+Consider the following tree structure where Document Type alias is presented in square brackets.
+
+* Codegarden
+  * 2023 \[_year_]
+    * Talks \[_talks_]
+      * ...
+      * Umbraco anno MMXXIII \[_talk_]
+    * Stages \[_stages_]
+      * Social Space \[_stage_]
+      * No 10 \[_stage_]
+      * No 16 \[_stage_]
+      * The Theatre \[_stage_]
+  * 2022 \[_year_]
+    * Talks \[_talks_]
+      * ...
+    * Stages \[_stages_]
+      * Main Stage \[_stage_]
+      * The Barn \[_stage_]
+      * The Theatre \[_stage_]
+
+Consider configuring a picker on the _talk_ Document Type to choose a stage of the talk. Here you only want to present the stages for the actual year. To do this, you need to choose the parent as origin.
+
+Imagine being on the `Umbraco anno MMXXIII` node. This means the collection of content keys passed into the first query step will only contain the `Talks` content node.
+
+* First you can then query for the nearest ancestors of type `year`. This means the `2023` will be returned.
+* Next, you can query for the nearest descendants of type `stages`.
```

Comment:
```suggestion
* First, query for the nearest ancestors of the type `year`. This will return `2023`.
* Next, query for the nearest descendants of type `stages`.
```